frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -832,6 +858,110 @@ pub mod pallet {
             Ok(().into())
         }
 
+        /// rewards are claimed by the staker on contract_id.
+        ///
+        /// era must be in the range `[current_era - history_depth; active_era)`.
+        ///
+        /// Any user can call this function.
+        #[pallet::weight(T::WeightInfo::payout_stakers_alive_staked(T::MaxStakings::get()))]
+        pub fn claim(
+            origin: OriginFor<T>,
+            contract_id: SmartContract<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            let claimer = ensure_signed(origin)?;
+
+            // check if this contract is registered
+            let developer = Self::registered_developer(&contract_id)
+                .ok_or(Error::<T>::ContractNotRegistered)?;
+
+            // double check if this is valid contract address
+            ensure!(
+                Self::is_contract_valid(&contract_id),
+                Error::<T>::ContractIsNotValid
+            );
+
+            // check if it was ever staked on this contract.
+            let last_staked =
+                Self::contract_last_staked(&contract_id).ok_or(Error::<T>::NothingToClaim)?;
+
+            // check if the contract is already claimed in this era
+            let current_era = Self::current_era().unwrap_or(Zero::zero());
+            let mut last_claim_era: EraIndex =
+                Self::contract_last_claimed(&contract_id).unwrap_or(current_era.clone());
+            ensure!(
+                current_era != last_claim_era,
+                Error::<T>::AlreadyClaimedInThisEra
+            );
+
+            // is this first time we claim the contract reward?
+            if last_claim_era.is_zero() {
+                last_claim_era = current_era.clone()
+            }
+
+            // oldest era to start with collecting rewards
+            let last_allowed_era = current_era.saturating_sub(Self::history_depth());
+            let start_from_era = last_claim_era.max(last_allowed_era);
+            if (start_from_era > last_claim_era) {
+                // TODO collect all unclaimed rewards and send to Treasury pallet
+            }
+
+            // for the first claimable era "start_from_era", this storage item must be in place!
+            let mut contract_stake_prev = Self::contract_era_stake(&contract_id, &start_from_era)
+                .ok_or(Error::<T>::UnknownStartStakingData)?;
```

Comment:
I don't think this is correct since you can have a situation like this:

Eras in which contract was staked:
1, 2, 3, 10, 11, 12, 25
Current era is: 30
History depth is: 10

Based on this, the only era staking points you have will be for era 25.
But this code will raise an error.

Reply:
We discussed this so I will mark it as resolved.

You will make a copy of the latest state and store it.